extra_geom/base.py

```diff
@@ -791,3 +798,135 @@ def offset(self, shift, *, modules=np.s_[:], tiles=np.s_[:]):
+
+        By default, this rotates all modules & tiles.
+        Returns a new geometry object of the same type.
+
+        ::
+
+            # Rotate the whole geometry by 90 degree in the xy plan
+            geom2 = geom.rotate((0, 0, 90))
+
+            # Move the tile 0 in the module 0 around its center by 90 degrees
+            geom2 = geom.rotate((0, 0, 90), modules=np.s_[:1], tiles=np.s_[:1])
+
+            # Rotate each module by a separate amount
+            rotate = np.zeros((16, 3))
+            rotate[5] = (3, 5, 1)  # x, y, z for individual modules
+            rotate[10] = (0, -2, 1)
+            geom2 = geom.rotate(rotate)
+
+        Parameters
+        ----------
+
+        angles: np.array or tuple
+            (x, y, z) rotations to apply in degree. Can be a single rotation
+            for all selected modules, a 2D array with a rotation per module, or
+            a 3D array with a rotation per tile (``arr[module, tile, xyz]``).
+        center: np.array or tuple
+            center of rotation. Shape must match angles.shape.
+            If set to None (default), the rotation center is set to:
+            * all modules: center of the detector
+            * selected modules: centers of modules
+            * selected tiles: centers of tiles
+        modules: slice
+            Select modules to rotate; defaults to all modules.
+        tiles: slice
+            Select tiles to move within each module; defaults to all tiles.
+        degrees: bool
+            If True (default), angles are in degrees. If False, angles are in
+            radians.
+        """
+        rot = np.asarray(angles)
+        if rot.shape[-1] != 3:
+            raise ValueError('Rotation angles must be 3d (x, y, z). '
+                             f'Last dimension was {rot.shape[-1]}')
+        if center is not None and np.asarray(center).shape != rot.shape:
+            raise ValueError('Rotation center must have same shape as angles')
+
+        ntiles = len(max((m for m in self.modules), key=len))
+        all_rot = np.zeros((len(self.modules), ntiles, 3), dtype=rot.dtype)
+        sel_rot = all_rot[modules, tiles]
+
+        refs = np.zeros((len(self.modules), ntiles, 3), dtype=np.float64)
+        sel_refs = refs[modules, tiles]
+
+        if center is None:
+            if rot.shape[:-1] == sel_rot.shape[:2] or tiles != np.s_[:]:
+                # Per-tile rotation
+                refs[modules, tiles] = np.array(
+                    [[t.centre() for t in m[tiles]]
+                    for m in self.modules[modules]]
+                )
+            elif rot.shape[:-1] == sel_rot.shape[:1] or modules != np.s_[:]:
+                # Per-module rotation
+                mods = []
+                for module in self.modules[modules]:
+                    center = np.mean([t.centre() for t in module], axis=0)
+                    mods.append(np.array([center for _ in module[tiles]]))
+                refs[modules, tiles] = np.array(mods)
```

Comment:
It took me a while to realise that we don't need a final `else:` here, because refs already contains zeros, which are the default when the whole detector is selected.

Reply:
Want me to add an empty branch with that comment, for clarity?

Reply:
Or just a comment where the `else` would go - whatever you think is clear.